### PR TITLE
Feat: Add Proof of Inclusion Plugin

### LINKED ISSUE #102: [Task]: Add Proof of Inclusion Plugin
### Task description

It would be super nice to add a button to the "block" view to "create proof of inclusion" for that block ID, in case the `plugins/poi/v1` is enabled.

By clicking that button, you send a get request to `api/plugins/poi/v1/create/{blockID}`.
You download the response as a json file.

Under Plugins Tab => "Proof of Inclusion", you add a mask to choose a file (which is the newly downloaded json file) and a "Validate" button. By clicking the button, you send a POST request with the file to `/api/plugins/poi/v1/validate`.

Then you show if the inclusion proof was valid or not.

### Requirements

-

### Acceptance criteria

-

### Creation checklist

- [ ] I have assigned this task to the correct people
- [ ] I have added the most appropriate labels
- [ ] I have linked the correct milestone and/or project

### LINKED ISSUE #225: [Task]: Implement PluginService in Participation and Spammer plugins
### Task description

Refactor Participation and Spammer plugins and implement Plugin Service to make plugin API calls similar to Poi plugin.

### Requirements

Plugin Service is implemented on branch created in this PR- https://github.com/iotaledger/node-dashboard/pull/224

### Acceptance criteria

Participation and Spammer plugins use Plugin Service

### Creation checklist

- [X] I have assigned this task to the correct people
- [X] I have added the most appropriate labels
- [X] I have linked the correct milestone and/or project

### PR DESCRIPTION
**Pull request by brancoder**
# Description of change

- Added tool to download PoI.json file on Block page
- Added Proof of inclusion plugin to validate PoI file
- Refactored Participation and Spammer plugins to utilise Plugin Service as used in Poi plugin.

Resolves https://github.com/iotaledger/node-dashboard/issues/102 .

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/app/components/plugins/Poi.tsx [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+import BlockButton from "../layout/BlockButton";
+import MilestonePayload from "../tangle/MilestonePayload";
+import TaggedDataPayload from "../tangle/TaggedDataPayload";
+import TransactionPayload from "../tangle/TransactionPayload";
+import { PoiState } from "./PoiState";
+
+/**
+ * Poi panel.
+ */
+class Poi extends AsyncComponent<unknown, PoiState> {
```
Comment (msarcev) [inline src/app/components/plugins/Poi.tsx]: Can we just name this component `ProofOfInclusion` ?
Having "Poi" written everywhere throughout code seams a bit cryptic, for no apparent reason.

File: src/services/pluginService.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+     */
+    private buildClient(): IClient {
+        const headers = this._authService.buildAuthHeaders();
+
+        return new SingleNodeClient(
+            `${window.location.protocol}//${window.location.host}`,
+            {
+                basePath: "/dashboard/api/",
+                headers
+            });
```
Comment (msarcev) [inline src/services/pluginService.ts]: Parenthesis goes to next line.

File: src/models/plugins/IValidateProofResponse.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface IValidateProofResponse {
```
Comment (msarcev) [inline src/models/plugins/IValidateProofResponse.ts]: Consider `IValidatePoiResponse`.

File: src/models/plugins/IPoi.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { IBlock, IMilestonePayload } from "@iota/iota.js";
+
+export interface IPoi {
```
Comment (msarcev) [inline src/models/plugins/IPoi.ts]: Seams better to spell the whole name here.

File: src/models/plugins/IPoi.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+    /**
+     * The milestone
+     */
+    milestone: IMilestonePayload;
+
+    /**
+     * The proof
+     */
+    proof: any;
```
Comment (msarcev) [inline src/models/plugins/IPoi.ts]: We don't know the type of this field ?

File: src/utils/downloadHelper.ts [hunk 6]
```diff
@@ -53,4 +53,25 @@ export class DownloadHelper {
         const b64 = Converter.bytesToBase64(Converter.utf8ToBytes(Converter.bytesToBase64(data)));
         return `data:plain/text;base64,${b64}`;
     }
+
+    /**
+     * Create a data url for base64 data.
```
Comment (msarcev) [inline src/utils/downloadHelper.ts]: Copypasta here.
Consider `Trigger a file download`.

File: src/app/components/plugins/Spammer.tsx [hunk 7]
```diff
@@ -201,71 +186,36 @@ class Spammer extends AsyncComponent<unknown, SpammerState> {
...
+                    isRunning: response.running ?? false,
+                    bps: response.bpsRateLimit?.toString() ?? "",
+                    cpu: (response.cpuMaxUsage ?? 0 * 100).toString(),
+                    workers: response.spammerWorkers?.toString() ?? "",
+                    workersMax: response.spammerWorkersMax ?? 0,
+                    valueSpamEnabled: response.valueSpamEnabled ?? false
                 });
             } else {
-                console.log("loging eror", response.error);
+                console.log("loging eror", response?.error);
```
Comment (msarcev) [inline src/app/components/plugins/Spammer.tsx]: "loging eror" seams useless, maybe "Error fetching spammer status" ?

File: src/app/components/plugins/PoiState.tsx [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { IPoi } from "../../../models/plugins/IPoi";
+
+export interface PoiState {
+    /**
+     * Is the dialog busy.
```
Comment (msarcev) [inline src/app/components/plugins/PoiState.tsx]: Fix jsdoc.

File: src/app/components/plugins/PoiState.tsx [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+     */
+    jsonData?: IPoi;
```
Comment (msarcev) [inline src/app/components/plugins/PoiState.tsx]: Consider calling it `poi` or `proofOfInclusion`.

File: src/app/components/plugins/Poi.tsx [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+import AsyncComponent from "../../components/layout/AsyncComponent";
+import Spinner from "../../components/layout/Spinner";
+import BlockButton from "../layout/BlockButton";
+import MilestonePayload from "../tangle/MilestonePayload";
+import TaggedDataPayload from "../tangle/TaggedDataPayload";
+import TransactionPayload from "../tangle/TransactionPayload";
+import { PoiState } from "./PoiState";
+
+/**
+ * Poi panel.
```
Comment (msarcev) [inline src/app/components/plugins/Poi.tsx]: Consider `Proof of Inclusion Plugin` here

**Review by msarcev — CHANGES_REQUESTED** [review]

**Comment by msarcev** [comment]
With node-docker-setup + local node-dashboard environment, when i click
image
on a block, there's an error in console
image
Is it supposed to work with this setup, or only with embedded dashboard ?

Navigating to that resource returns:
`{"error":{"code":"400","message":"missing or malformed jwt, error: code=400, message=missing or malformed jwt"}}`

Maybe i skipped some JWT configuration somewhere ? 
But i'm logged into dashboard so i guess it should have a JWT.

File: src/models/plugins/IValidateProofResponse.ts (hunk 3 above)
Comment (brancoder) [reply src/models/plugins/IValidateProofResponse.ts]: Named it IValidateProofResponse to match the inx-poi model but makes more sense to change ti to IValidatePoiResponse be consistent.

**Comment by brancoder** [comment]
> 

Tha lates inx-dashboard doesnt have poi routes exposed and that will be fixed in the next release. You can test using the branch [feat/popi](https://github.com/iotaledger/inx-dashboard/tree/feat/poi).

**Pull request merged**